system/middleware.go

```diff
@@ -50,33 +48,34 @@ func (application *Application) ApplyDbMap(c *web.C, h http.Handler) http.Handle
 func (application *Application) ApplyAPI(c *web.C, h http.Handler) http.Handler {
 	fn := func(w http.ResponseWriter, r *http.Request) {
 		if strings.HasPrefix(r.URL.Path, "/api") {
+			var user *models.User
+			var err error
+			dbMap := c.Env["DbMap"].(*gorp.DbMap)
+
 			authHeader := r.Header.Get("Authorization")
 			if strings.HasPrefix(authHeader, "Bearer ") {
-				apitoken := strings.TrimPrefix(authHeader, "Bearer ")
-
-				JWTtoken, err := jwt.Parse(apitoken, func(token *jwt.Token) (interface{}, error) {
-					// validate signing algorithm
-					if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
-						return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
-					}
-					return []byte(application.APISecret), nil
-				})
-
-				if err != nil {
-					log.Warnf("invalid token %v: %v", apitoken, err)
-				} else if claims, ok := JWTtoken.Claims.(jwt.MapClaims); ok && JWTtoken.Valid {
-					dbMap := c.Env["DbMap"].(*gorp.DbMap)
-
-					user, err := models.GetUserByID(dbMap, int64(claims["loggedInAs"].(float64)))
-					if err != nil {
-						log.Errorf("unable to map apitoken %v to user id %v", apitoken, claims["loggedInAs"])
-					} else {
-						c.Env["APIUserID"] = user.ID
-						log.Infof("mapped apitoken %v to user id %v", apitoken, user.ID)
-					}
+				var userId int64
+				userId, err = application.validateToken(authHeader)
+				if err == nil {
+					user, err = models.GetUserByID(dbMap, userId)
 				}
+			} else if strings.HasPrefix(authHeader, "TicketAuth ") {
+				var userMsa string
+				userMsa, err = application.validateTicketOwnership(authHeader)
+				if err == nil {
+					user, err = models.GetUserByMSA(dbMap, userMsa)
+				}
+			}
+
+			if err != nil {
+				log.Warnf("api authorization failure: %v", err)
+				c.Env["AuthErrorMessage"] = err.Error()
+			} else {
+				c.Env["APIUserID"] = user.ID
+				log.Infof("mapped api auth header %v to user %v", authHeader, user.ID)
 			}
 		}
+
```

Comment:
Pretty sure the panic is here. In the case where the header is not "Bearer " or "TicketAuth " error will be nil, causeing a panic when there is no user.ID to set and log at the end.

Reply:
Yeah, that explains the panic you reported earlier. Fixed.